frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2674,131 +2681,137 @@ std::vector<ID> findUsedImportedIds(Context* context,
-static void findAllModulesUsedImportedInTreeImpl(Context* context,
-                                                 std::map<ID, ID>& ret,
-                                                 ID id) {
-  if (id.isEmpty()) return;
+struct GatherUsedImportedIds {
+  Context* context;
+  const Module* inMod = nullptr;
+  std::set<const Scope*> scopes;
+  std::set<ID> idSet;
+  std::vector<ID> idVec;
 
-  auto ast = parsing::idToAst(context, id);
+  GatherUsedImportedIds(Context* context, const Module* inMod)
+    : context(context), inMod(inMod)
+  { }
 
-  if (auto scope = scopeForId(context, id)) {
-    if (auto rvs = resolveVisibilityStmts(context, scope)) {
-      for (auto& vc : rvs->visibilityClauses()) {
-        auto& k = vc.visibilityClauseId();
-        auto& v = vc.scope()->id();
-        ret.insert(std::make_pair(k, v));
+  void processUseImport(const AstNode* ast) {
+    const Scope* scope = scopeForId(context, ast->id());
+    if (scope) {
+      auto p = scopes.insert(scope);
+      if (p.second) {
+        // Insertion occured, so this is the first time visiting this scope.
+        // Gather the IDs of the used/imported modules
+        auto vec = findUsedImportedIds(context, scope);
+        for (const auto& id: vec) {
+          // filter out enums since they are not interesting here
+          if (parsing::idIsModule(context, id)) {
+            // save the module used/imported to ids
+            auto p2 = idSet.insert(id);
+            if (p2.second) {
+              // insertion occured, so add it also to the vector
+              idVec.push_back(id);
+            }
+          }
+        }
       }
     }
   }
 
-  // Recurse, but skip modules, since we do not consider their use/imports.
-  for (auto child : ast->children()) {
-    if (child->isModule()) continue;
-    findAllModulesUsedImportedInTreeImpl(context, ret, child->id());
+  // gather information about what use and import statements refer to
+  bool enter(const Use* d) {
+    processUseImport(d);
+    return false;
+  }
+  void exit(const Use* d) { }
+  bool enter(const Import* d) {
+    processUseImport(d);
+    return false;
+  }
+  void exit(const Import* d) { }
+
+  // do not delve into submodules
+  bool enter(const Module* m) {
+    return (m == inMod); // visit the requested module only
+  }
+  void exit(const Module* m) { }
+
+  // traverse through anything else
+  bool enter(const AstNode* ast) {
+    return true;
+  }
+  void exit(const AstNode* ast) { }
+};
+
+static const std::vector<ID>&
+findAllModulesUsedImportedInTreeQuery(Context* context, ID modId) {
+  QUERY_BEGIN(findAllModulesUsedImportedInTreeQuery, context, modId);
+  std::vector<ID> ret;
+
+  auto ast = parsing::idToAst(context, modId);
+  if (ast) {
+    if (auto mod = ast->toModule()) {
+      GatherUsedImportedIds gatherModules(context, mod);
+      ast->traverse(gatherModules);
+      // swap the gathered vector in to place
+      ret.swap(gatherModules.idVec);
+    }
   }
```

Comment:
If you care about query projection-ness (it kinda sounded like you did in prior PRs?), wouldn't `idToTag(context, modId) == MODULE` here, followed by an (unchecked) `idToAst(...)->toModule()` improve memoization behavior?

Reply:
I do care about query projection-ness but in this case, it's a misuse of this function if it's not a module. I'll adjust it to use CHPL_ASSERT to make this clearer.